crates/sui-core/src/quorum_driver/mod.rs

```diff
@@ -763,9 +792,23 @@ where
     ) {
         let tx_digest = *transaction.digest();
         if let Some(qd_error) = err {
-            debug!(?tx_digest, "Failed to {action}: {}", qd_error);
-            // non-retryable failure, this task reaches terminal state for now, notify waiter.
-            quorum_driver.notify(&transaction, &Err(qd_error), old_retry_times + 1);
+            match qd_error {
+                // Special case for SystemOverloadRetryAfter error. In this case, due to that objects are already
+                // locked inside validators, we need to perform continuous retry and ignore `max_retry_times`.
+                QuorumDriverError::SystemOverloadRetryAfter { .. } => {
+                    debug!(?tx_digest, "Failed to {action} - Retrying");
+                    spawn_monitored_task!(quorum_driver.backoff_and_enqueue(
+                        transaction.clone(),
+                        tx_cert,
+                        old_retry_times
```

Comment:
Could retry times overflow now that this can go on for a while?

Reply:
good point - there's no need for it to be a u8

Reply:
Good point. I added a TODO for only allowing limited number of active transactions in a quorum driver. When the limit is reached, the quorum driver should reject any new transaction requests.